File: 189-rotate-array/189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int l = nums.length;
        int[] temp = new int[l];
        k%=l;
        
        for(int i=0;i<k;i++){
            temp[i] = nums[l-k+i];
        }
        
        int i=0;
        for(int j=k;j<l;j++){
            temp[j] = nums[i++];
        }
        
        for(int x=0;x<l;x++){
            nums[x] = temp[x];
        }
        
    }
}

File: 189-rotate-array/189-rotate-array.cpp
class Solution {
public:
    
    void rotate(vector<int>& nums, int k) {
         //if k==array.size() then no change after rotation
        if(k == nums.size()) return;
        else k = k%nums.size();
           
        //step 1: reversing first n-k elements, array:[3, 2, 1, 4, 5]
        reverse(nums.begin(),nums.end()-k);
           
        //step 2: reversing remaining k elements, array:[3, 2, 1, 5, 4]
        reverse(nums.begin()+nums.size()-k, nums.end());
           
        //step 3: reversing whole vector, array:[4, 5, 1, 2, 3], this is resultant
        reverse(nums.begin(), nums.end());
    }
};